File: go/src/filter.go
package main

import (
	"errors"
	"fmt"
	"strings"

	"github.com/go-pg/pg/v10"
	"github.com/go-pg/pg/v10/orm"
	"github.com/mozillazg/go-unidecode"
)

type Sort struct {
	Sort  string `json:"sort"`
	ColId string `json:"colId"`
}

type Filter struct {
	FilterType string `json:"filterType"`
	Type       string `json:"type"`
	Filter     string `json:"filter"`
}

type GridRequest struct {
	Start        int               `json:"start"`
	End          int               `json:"end"`
	Sort         []Sort            `json:"sort"`
	Filter       map[string]Filter `json:"filter"`
	GlobalSearch string            `json:"globalSearch"`
}

func removeAccents(s string) string {
	return strings.ToLower(unidecode.Unidecode(s))
}

func filterSortAndPage(dataset *orm.Query, request GridRequest) (*orm.Query, error) {
	// Filter
	for key, filter := range request.Filter {
		if filter.FilterType != "text" {
			return nil, errors.New("Unsupported filter type")
		}

		col := pg.Ident(key)
		unaccentedFilter := removeAccents(filter.Filter)

		if filter.Type == "equals" {
			dataset = dataset.Where("unaccent(?) = ?", col, unaccentedFilter)
		} else if filter.Type == "notEquals" {
			dataset = dataset.Where("unaccent(?) <> ?", col, unaccentedFilter)
		} else if filter.Type == "contains" {
			dataset = dataset.Where("unaccent(?) ilike ?", col, "%"+unaccentedFilter+"%")
		} else if filter.Type == "notContains" {
			dataset = dataset.Where("NOT unaccent(?) ilike ?", col, "%"+unaccentedFilter+"%")
		} else if filter.Type == "startsWith" {
			dataset = dataset.Where("unaccent(?) ilike ?", col, unaccentedFilter+"%")
		} else if filter.Type == "endsWith" {
			dataset = dataset.Where("unaccent(?) ilike ?", col, "%"+unaccentedFilter)
		} else if filter.Type == "blank" {
			dataset = dataset.Where("(? <> '') IS NOT TRUE", col)
		} else if filter.Type == "notBlank" {
			dataset = dataset.Where("? <> ''", col)
		} else {
			return nil, errors.New("Unsupported type")
		}
	}

	// Global filter
	if len(request.GlobalSearch) >= 0 {
		unaccentedGlobalSearch := removeAccents(request.GlobalSearch)

		dataset = dataset.WhereGroup(func(q *pg.Query) (*pg.Query, error) {
			for _, field := range globalSearchableFields {
				q = q.WhereOr("unaccent(?) ilike ?", pg.Ident(field), "%"+unaccentedGlobalSearch+"%")
			}
			return q, nil
		})
	}

	// Sort
	if len(request.Sort) > 0 {
		sort := request.Sort[0]
		dataset = dataset.Order(fmt.Sprintf("%s %s", sort.ColId, sort.Sort))
	}

	// Page
	return dataset.Offset(request.Start).Limit(request.End - request.Start), nil
}

File: cpp/src/filter.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <list>

#include <pqxx/pqxx>

#include "../include/crow_all.h"
#include "../include/utils.hpp"
#include "../include/filter.hpp"

using namespace std;
using namespace pqxx;

bool filter_sort_and_page(stringstream& query, params& prms, crow::json::rvalue& request, work& tx, list<string>& global_searchable_fields) {
    int start(request["start"]);
    int end(request["end"]);
    std::string global_search(request["globalSearch"]);

    query << " WHERE 1 = 1 ";

    if (!global_search.empty()) {
        query << "AND (1 = 0 ";

        for (auto& field: global_searchable_fields) {
            query << "OR position($" << prms.size() + 1 << " in " << tx.esc(field) << ") > 0 ";
        }

        query << ") ";
        prms.append(global_search);
    }

    for(auto& el: request["filter"].keys()) {
        string field = el;
        auto& filter_value = request["filter"][field];

        string filter(filter_value["filter"]);
        string filter_type(filter_value["filterType"]);
        string type(filter_value["type"]);
        auto col = tx.esc(camel_to_snake(field));

        if (!is_valid_col_name(col)) {
            cerr << "Invalid column name: " << col << endl;
            return false;
        }

        if (filter_type != "text") {
            cerr << "Unsupported filter type" << endl;
            return false;
        }

        if (type == "equals") {
            query << "AND (" << col << " = $" << prms.size() + 1 << ") ";
        } else if (type == "notEquals") {
            query << "AND (" << col << " <> $" << prms.size() + 1 << ") ";
        } else if (type == "contains") {
            query << "AND (position($" << prms.size() + 1 << " in " << col << ") > 0) ";
        } else if (type == "notContains") {
            query << "AND (position($" << prms.size() + 1 << " in " << col << ") = 0) ";
        } else if (type == "startsWith") {
            query << "AND (" << col << " LIKE $" << prms.size() + 1 << ") ";
            filter = filter + '%';
        } else if (type == "endsWith") {
            query << "AND (" << col << " LIKE $" << prms.size() + 1 << ") ";
            filter = '%' + filter;
        } else if (type == "blank") {
            query << "AND (" << col << " <> '') IS NOT TRUE ";
        } else if (type == "notBlank") {
            query << "AND (" << col << " <> '') ";
        } else {
            cerr << "Unsupported type" << endl;
            return false;
        }

        if (type != "blank" && type != "notBlank") {
            prms.append(filter);
        }
    }

    auto sort = request["sort"];

    if (sort.size() > 0) {
        auto first = sort[0];
        string col_id(first["colId"]);
        string sort_with_case(first["sort"]);
        string sort_lower(to_lower_case(sort_with_case));

        auto col_snake = tx.esc(camel_to_snake(col_id));

        if (!is_valid_col_name(col_snake)) {
            cerr << "Invalid column name: " << col_snake << endl;
            return false;
        }

        if (sort_lower != "asc" && sort_lower != "desc") {
            cerr << "Invalid sort direction" << endl;
            return false;
        }

        // We shouldn't be able to inject anything since only a-z and _ are allowed
        query << "ORDER BY " << col_snake << " " << tx.esc(sort_lower) << " ";
    }

    query << "OFFSET $" << prms.size() + 1 << " "
          << "LIMIT $" << prms.size() + 2 << " ";

    prms.append(to_string(start));
    prms.append(to_string(end - start));

    return true;
}
